tests/Unit/Emogrifier/HtmlProcessor/CssInlinerTest.php

```diff
@@ -1281,6 +1281,87 @@ public function emogrifyMatchesRuleAfterUnneededCssThing($css)
+     * that allows for such whitespace removal, so that the tests themselves do not need to be written less humanly
+     * readable and can use inputs containing extra whitespace.
+     *
+     * @param string $needle Needle that would be used with `assertContains` or `assertNotContains`
+     *
+     * @return string Needle to use with `assertRegExp` or `assertNotRegExp` instead
+     */
+    private static function getNeedleRegExpAllowingRemovalOfUnnecessaryCssWhitespace($needle)
+    {
+        $needleMatcher = preg_replace_callback(
+            '/\\s*+([{}])\\s*+|(?:(?!\\s*+[{}]).)++/',
+            function ($matches) {
+                if (!empty($matches[1])) {
+                    // matched possibly some whitespace, followed by "{" or "}", then possibly more whitespace
+                    return '\\s*+' . preg_quote($matches[1], '/') . '\\s*+';
+                } else {
+                    // matched any other sequence which could not overlap with the above
+                    return preg_quote($matches[0], '/');
+                }
+            },
+            $needle
+        );
+        return '/' . $needleMatcher . '/';
+    }
+
+    /**
+     * Like `assertContains` but allows for removal of some unnecessary whitespace from the CSS
+     *
+     * @param string $needle
+     * @param string $haystack
+     */
+    private static function assertContainsAllowingRemovalOfUnnecessaryCssWhitespace($needle, $haystack)
+    {
+        static::assertRegExp(
+            static::getNeedleRegExpAllowingRemovalOfUnnecessaryCssWhitespace($needle),
+            $haystack,
+            'Plain text needle: "' . $needle . '"'
+        );
+    }
+
+    /**
+     * Like `assertNotContains` and also enforces the assertion with removal of some unnecessary whitespace from the CSS
+     *
+     * @param string $needle
+     * @param string $haystack
+     */
+    private static function assertNotContainsAllowingRemovalOfUnnecessaryCssWhitespace($needle, $haystack)
+    {
+        static::assertNotRegExp(
+            static::getNeedleRegExpAllowingRemovalOfUnnecessaryCssWhitespace($needle),
+            $haystack,
+            'Plain text needle: "' . $needle . '"'
+        );
+    }
+
+    /**
+     * Asserts that a string of CSS occurs exactly a certain number of times in the result, allowing for removal of some
+     * unnecessary whitespace
+     *
+     * @param string $needle
+     * @param int $expectedCount
+     * @param string $haystack
+     */
+    private static function assertContainsCountAllowingRemovalOfUnnecessaryCssWhitespace(
+        $needle,
+        $expectedCount,
+        $haystack
+    ) {
+        static::assertSame(
+            $expectedCount,
+            preg_match_all(static::getNeedleRegExpAllowingRemovalOfUnnecessaryCssWhitespace($needle), $haystack),
+            'Plain text needle: "' . $needle . "\"\nHaystack: \"" . $haystack . '"'
```

Comment:
If you use single quotes for the middle string, you won't need to escape the double quotes.

Reply:
But then to have the newline would require concatenating three strings:
`'"' . "\n" . 'Haystack: "'`

Reply:
Oh, right. I missed that.